### PR TITLE
Add support for edge reconnection

### LINKED ISSUE #780: Reconnecting edges
* [x] **I have checked that this feature has not yet been suggested by someone else.**

### Expected behavior

Add a way to reconnect existing edges to match parity with Sirius Desktop.

image

It would be great if _Reconnect Edge_ tools were taken into account too.

### Cuttings backs

I don't know what this section means.

### Potential side effects

Not sure.

### Describe alternatives you've considered

Users can delete the edge and create a new one, or change source/target in the _Details_ pane. The former solution results in subpar UX, the latter does not leverage the UI and requires picking an element from a potentially long list.

### PR DESCRIPTION
**Pull request by gcoutable**
# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [x] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [x] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [x] Have the relevant issues been added to the pull request?
- [x] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [x] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [x] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [x] Variables have a proper type
- [x] Functions’ arguments have a proper type
- [x] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

**Comment by gcoutable** [comment]
This PR is ready for review but since it needs both #997 and #931 to work, I keep this PR in draft.

File: CHANGELOG.adoc [hunk 1]
```diff
@@ -33,6 +34,7 @@
 
 - https://github.com/eclipse-sirius/sirius-components/issues/1316[#1316] [diagram] Support parametric svg node style
 - https://github.com/eclipse-sirius/sirius-components/issues/997[#997] [diagram] Add support for edition of routing points in diagrams
+- https://github.com/eclipse-sirius/sirius-components/issues/780[#780] [diagram] Add support for edge reconnection. Edge reconnection descriptions can be defined either inside a classical Sirius desktop VSM or, using the View DSL. While using the Sirius desktop VSM you can defined a tool to do both target and source reconnection, you will need to defined a tool to reconnect the source and another one to reconnect the target using the View DSL.
```
Comment (pcdavid) [inline CHANGELOG.adoc]: Typos:
- "you can defined" => "you can define"
- "need to defined" => "need to define"

File: doc/adrs/074_add_support_for_edge_reconnection.adoc [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+== Decision
+
+=== GraphQL API Change
+
+```
+extend type Mutation {
+  reconnectEdge(input: ReconnectEdgeInput!): ReconnectEdgePayload!
+}
+
+input ReconnectEdgeInput {
```
Comment (pcdavid) [inline doc/adrs/074_add_support_for_edge_reconnection.adoc:21]: This does not match what is actually in `diagram.graphqls`:

```
enum ReconnectEdgeKind {
  SOURCE
  TARGET
}

input ReconnectEdgeInput {
  id: ID!
  editingContextId: ID!
  representationId: ID!
  edgeId: String!
  newEdgeEndId: String!
  reconnectEdgeKind: ReconnectEdgeKind!
  newEdgeEndPosition: PositionInput!
}
```

File: doc/adrs/074_add_support_for_edge_reconnection.adoc [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  newEdgeEndPosition: InputPosition!
+}
+```
+
+We will leverage of the subscription to send the new diagram state to the frontend.
+
+=== Frontend
+
+We will rely as much as possible on Sprotty behavior.
+We will have to adapte the reconnect action to take account of the source and target anchor relative position ratio.
```
Comment (pcdavid) [inline doc/adrs/074_add_support_for_edge_reconnection.adoc]: Typo: "adapte" => "adapt"

File: doc/adrs/074_add_support_for_edge_reconnection.adoc [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+We will need to prevent Sprotty to reset routing point during the reconnect by preventing the `SiriusPolylineEdgeRouter` to call `resetRoutingPointsOnReconnect`.
+When the end is dropped, Sprotty tries to find the right node to reconnect the edge. Unfortunately Sprotty select the most distant ancestor on which the edge end is dropped while we wanted the element on which the edge end is dropped. To fix that we will override the `MoveMouseListener#mouseUp` to always take the node right under the drop position.
```
Comment (pcdavid) [inline doc/adrs/074_add_support_for_edge_reconnection.adoc]: - "Sprotty select" => "Sprotty selects"
- "we wanted" => "we want"
- missing line breaks between sentences

File: doc/adrs/074_add_support_for_edge_reconnection.adoc [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+We will need to improve the contract of the anchor computer since it should know where the anchor is and how to update the anchor position.
+We will need to override the sprotty reconnect action to makes it provide the drop position to send it the the backend. We will need to support that custom reconnect action in our diagram server. The mutation will be send only if the source or the target of the reconnecting edge has changed.
+To tell the user that the reconnection he tried is not possible we will leverage on the Snackbar to display the error return by the reconnect edge mutation.
+
```
Comment (pcdavid) [inline doc/adrs/074_add_support_for_edge_reconnection.adoc:47]: These are all very low-level technical/implementation details, but it's not clear how the overall feature works from the end-user's point of view:
- how does the visual feeback (if any) look like during a reconnect?
- how does the user know if the current mouse position is a valid reconnection end?
- what if the user tries to reconnect an edge that does _not_ have a reconnection tool?
- when reconnecting an edge which has multiple bendpoints, how are they affected? is the edge "straightened" and all bendpoints removed (below you say "reconnecting an edge is similar to removing an edge, then creating a new one", which would behave like that), or are the existing bendpoints preserved and only the first/segment modified?

File: packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/description/DiagramDescription.java [hunk 6]
```diff
@@ -104,6 +107,10 @@ public List<EdgeDescription> getEdgeDescriptions() {
         return this.edgeDescriptions;
     }
 
+    public List<ITool> getTools() {
+        return this.tools;
+    }
```
Comment (sbegaudeau) [inline packages/diagrams/backend/sirius-components-diagrams/src/main/java/org/eclipse/sirius/components/diagrams/description/DiagramDescription.java:112]: The `DiagramDescription` is not to be used as a delivery boy anymore than it is already. If you need to ask for a reconnection tool in the `ViewReconnectionToolsExecutor` then you should figure it out there. Don't add this data structure here only as a shortcut for the View DSL support.

**Review by sbegaudeau — CHANGES_REQUESTED** [review]
I haven't finish the review but here's a first comment.

**Review by sbegaudeau — APPROVED** [review]

**Pull request merged**